### ISSUE TITLE
Loading ONNX model

### ISSUE THREAD

**Issue by MotoCommando** [deeplearning4j/deeplearning4j#9133]
Hi, 
I'm interested in using deeplearning4j with my trained onnx model, however I could not find any reader for onnx format.I found in your documentation Conv2d class that uses Onnx protobuf classes ([https://deeplearning4j.org/api/latest/org/nd4j/linalg/api/ops/impl/layers/convolution/Conv2D.html]
Is there a way to load onnx model or is it still work in progress and how to use this class?

**Comment by agibsonccc** [comment]
@MotoCommando we are in the middle of an about finished rewrite of the model import framework. It should be done in a few weeks. For now, the second best option is to look in to converting it to tensorflow and loading it via model import there.

**Comment by MotoCommando** [comment]
Ok thanks for reply!
So I guess this Conv2D class is unusable for the time being?
Also is there a way to create convolution layers with predefined weights?

**Comment by agibsonccc** [comment]
@MotoCommando  that wasn't really the intended takeaway, Converting a model should only be a few lines of code: https://github.com/onnx/tensorflow-onnx/ the tensorflow import is fairly mature and should suit your needs (especially for simpler models) - again we'll have a new model import framework soon that supports multiple frameworks, but I imagine there are plenty of workarounds if you want to run in dl4j.

**Comment by MotoCommando** [comment]
@agibsonccc ok thanks for help. Maybe one more questions. Will Onnx also be stored in SameDiff class?

**Comment by agibsonccc** [comment]
@MotoCommando here's the PR that will address this. https://github.com/eclipse/deeplearning4j/pull/9165 
We'll be releasing a release candidate with the new framework soon. This will also replace our existing tensorflow import. I have a lot of docs to write but here's where it lives:
https://github.com/eclipse/deeplearning4j/tree/KonduitAI-master/nd4j/samediff-import/samediff-import-onnx
Here's what it looks like:
https://github.com/eclipse/deeplearning4j/blob/KonduitAI-master/nd4j/samediff-import/samediff-import-onnx/src/test/kotlin/org/nd4j/samediff/frameworkimport/onnx/importer/TestOnnxFrameworkImporter.kt#L27

Underneath the covers, we define a set of rules that exported and loaded as protobuf. The default rules are defined in kotlin and can be found here:
https://github.com/eclipse/deeplearning4j/blob/KonduitAI-master/nd4j/samediff-import/samediff-import-onnx/src/main/kotlin/org/nd4j/samediff/frameworkimport/onnx/definitions/OnnxOpDeclarations.kt

The op definitions for nd4j can be found here:
https://github.com/eclipse/deeplearning4j/blob/KonduitAI-master/nd4j/samediff-import/samediff-import-api/src/main/resources/nd4j-op-def.pbtxt 

As you can see, nd4j actually supports quite a few ops and various parameters. It just wasn't surfaced well, I'm hoping we'll be able to document these things better going forward with automation.

If you find a bug, you can actually override the file used allowing for fairly easy migration of model import to a fixed version. 
We will also allow pre processing in the next version for ops we don't support directly which will allow import with equivalent results. You can see an example of what that looks like here:
https://github.com/eclipse/deeplearning4j/blob/KonduitAI-master/nd4j/samediff-import/samediff-import-onnx/src/test/kotlin/org/nd4j/samediff/frameworkimport/onnx/processing/GroupConvPreProcessingRule.kt



Basically we'll allow pre and post processors to be defined using annotations. More soon.

**Comment by DarkWingMcQuack** [comment]
Are there any news on importing ONNX models directly? Is there a tutorial or example code how to do this?

**Comment by agibsonccc** [comment]
@DarkWingMcQuack  yes you can use the samediff-import-onnx. It's not fully baked yet though. If you want to give it a shot, I'd be happy to take a look at your model and see if it works.
Basic docs for that are here: https://deeplearning4j.konduit.ai/v/en-1.0.0-m1/samediff/explanation/model-import-framework 

You can see the implemented definitions here:
https://github.com/eclipse/deeplearning4j/blob/master/nd4j/samediff-import/samediff-import-onnx/src/main/kotlin/org/nd4j/samediff/frameworkimport/onnx/definitions/OnnxOpDeclarations.kt

 Otherwise, you can just use the nd4j-onnxruntime for now: https://deeplearning4j.konduit.ai/v/en-1.0.0-m1/nd4j/how-to-guides/other-framework-interop/onnx

**Comment by agibsonccc** [comment]
Closing. We support quite a few models now. For anyone finding this in the future file an issue if we're missing something.
